### PR TITLE
feat: refactor sig_circuit and ecc_circuit row capacity estimation

### LINKED ISSUE #648: change degree of sig circuit to 20; impl min_num_rows_block for sig circuit / ecc circuit
before start working on this issue we need first solve https://scrollco.slack.com/archives/C02PP3NNDC5/p1689675472809549

other circuits except sig circuit / ecc circuit, they are all "row by row" circuit. Each row indicates some input data, we constrains gates for each row (if related selectors enabled). But sig circuit / ecc circuit (halo2-lib based), they are all vertical circuit. A lot of `enable_equal` + standard gate "flex_gate" used to describe the computation constraints.

For for those vertical circuits, behavior of min_num_rows_block should be carefully designed.

Currently, we use `min_num_rows_block` inside l2geth to decide whether a block is full for circuit row capacity.  So i would propose,   
(1) for sig circuit, we know cells for each sig verification, so what about make min_num_rows_block return NUM_TX * cells / columns?   So if circuit degree = 20 (2\*\*20 \~= 1M), and  SigCircuit::min_num_rows_block return 0.5M, we know "half of sig circuit capacity" is used, if  SigCircuit::min_num_rows_block return 1.5M, we know oh it overflows. Although this is not how row is really used inside sig circuit, it is more consistent with how we handle other circuits.  ( i guess some code refactor needed? we should separate what is static (circuit config level) and what is runtime (block witness level) better for better understanding.  btw currently the default DEGREE of sig circuit is 19 (prod circuit degree is 20, using 19 means half cells wasted), , i remember it is due to some historical reason mentioned above in slack.  We should change it to 20, (something like 1<<20 - 256 ?).     
(2) for ecc circuit, for the block with num_ec_add, num_ec_mul, num_ec_pairing,  let min_num_rows_block to return max_of(2\*\*20 / max_num_ec_add * num_ec_add, 2\*\*20 / max_num_ec_mul * num_ec_mul, 2\*\*20 / max_num_ec_pairing * num_ec_pairing)?    If if circuit overflows, log::error(num_ec_add, num_ec_mul, num_ec_pairing).

### PR DESCRIPTION
**Pull request by darth-cy**
### Description

1. Double row usage for sig circuit while avoiding blinding area in integration.
2. Implement and unify with other sub circuits the `min_num_rows_block` interface for ecc/sig circuits, both using`halo2-lib` vertical gate assignments.

### Issue Link

Closes #648

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: zkevm-circuits/src/ecc_circuit/util.rs [hunk 1]
```diff
@@ -2,8 +2,17 @@ use eth_types::Field;
...
 use halo2_ecc::{bigint::CRTInteger, ecc::EcPoint, fields::FieldExtPoint};
 
+// Total number of rows allowable for ECC circuit
 pub const LOG_TOTAL_NUM_ROWS: u32 = 20;
 
+// Cell usage accounting for EcAdd, EcMul and EcPairing
+// Roud up to nearest 100
+pub(super) const EC_ADD_CELLS: usize = 6_900;    // actual: 6_851
+pub(super) const EC_MUL_CELLS: usize = 405_500;  // actual: 405_476
+pub(super) const EC_PAIRING_CELLS: usize = 6_627_500; // actual: 6_627_442
```
Comment (darth-cy) [inline zkevm-circuits/src/ecc_circuit/util.rs]: To observe actual cell usage, trace-level log of ctx object's cursor position is added to allow direct row accounting. To observe cell usage for ECC, run:
`RUST_BACKTRACE=1 RUST_LOG=trace cargo test -p zkevm-circuits --features test ecc_circuit::test::test_ecc_circuit`

> Sample Output:
```
...
> [ECC] ==> EcAdd Assignmnet START:
> Ctx cell pos: [(0, 0)]
> [ECC] EcAdd Inputs Assigned:
> Ctx cell pos: [(0, 1014)]
> [ECC] EcAdd Assignmnet END:
> Ctx cell pos: [(0, 6851)]
...
> [ECC] ==> EcMul Assignmnet START:
> Ctx cell pos: [(0, 342550)]
> [ECC] EcMul Inputs Assigned:
> Ctx cell pos: [(0, 343298)]
> [ECC] EcMul Assignmnet END:
> Ctx cell pos: [(0, 748027)]
...
> [ECC] ==> EcPairing Assignment START:
> Ctx cell pos: [(19, 693721)]
> [ECC] EcPairing g1s Assigned:
> Ctx cell pos: [(19, 695073)]
> [ECC] EcPairing g2s Assigned:
> Ctx cell pos: [(19, 697777)]
> [ECC] EcPairing Inputs RLC Assigned:
> Ctx cell pos: [(19, 700082)]
> [ECC] EcPairingAssignment END:
> Ctx cell pos: [(25, 1029779)]
...
```
The `max_rows` on ctx already takes into account the blinding area (in this case 9 rows). `max_rows` in this test is 1048567 (= 2^20 - 9).

File: zkevm-circuits/src/ecc_circuit/util.rs [hunk 2]
```diff
@@ -2,8 +2,17 @@ use eth_types::Field;
 use halo2_base::{AssignedValue, QuantumCell};
 use halo2_ecc::{bigint::CRTInteger, ecc::EcPoint, fields::FieldExtPoint};
 
+// Total number of rows allowable for ECC circuit
```
Comment (roynalnaruto) [inline zkevm-circuits/src/ecc_circuit/util.rs:5]: Total number of rows will be `1 << 20`

**Comment by roynalnaruto** [comment]
@darth-cy I think we should also have a util method to compute the number of advices `num_advice` depending on the number of `EcAdd`, `EcMul` and `EcPairing` ops. These config values can be moved out of `EccCircuit` struct and into the `EccCircuitConfig` struct. cc @lispc

**Comment by darth-cy** [comment]
Agree with above. The `num_advise` can be dynamically configured for the field chip in `EccCircuitConfig`.

**Comment by lispc** [comment]
the second elem of the pair returned by min_num_rows is the "actual row num used by the subcircuit".  Eg: for sub circuits , the second elem usually is the row num counting padding rows., while the first elem of the returned pair is a "how many rows the subcircuit needs for the inputs minimally", usually row num without padding rows.   For example, for state circuit, the returned pair is (rws.len(), params.max_rws).   

 The first elem is used to decide when l2geth cannot add new tx into a block.   Simplified l2geth code

``` l2geth code
let row_usage_of_each_subcircuit = [0,0,0,...];
while true:
    tx = pick_new_tx_from_mempool();
    tx_trace = exec_tx(tx)
    tx_row_usage = SuperCircuit::min_num_rows_block_subcircuits(tx_trace).map(|pair| pair.0).collect_vec();
    row_usage_of_each_subcircuit += tx_row_usage; // element wise
    if max_of(row_usage_of_each_subcircuit) > 1<<20:
       // it is time to build a block
       roll_back_last_tx
       seal block
       reset row_usage_of_each_subcircuit 
       continue
```

 For the halo2-base circuits, the second elem returned is correct.  But for first elem, i would suggest another way: Say the sig circuit is 50 columns, degree = 20, and can includes 100 sigs max. Then for blocks with 10 sigs, actually the sig circuit needs only 5 columns (the other 45 columns are used to compute padded sigs).  So for this case, we return 1<<20 / 100 \* 5 as the first elem of min_num_rows.   For a block with 120 sigs, we return 1<<20 / 100 \* 120 as the first elem of min_num_rows.  So the first elem is no longer real num usage, but just an indicator of circuit capacity usage ratio. It is a metric that can be added. I explained this in the issue descritpion too, but not very clearly.

**Comment by darth-cy** [comment]
Just saw the last comment. Will modify the impl accordingly. This explanation is very clear!

File: zkevm-circuits/src/sig_circuit.rs [hunk 3]
```diff
@@ -235,8 +253,21 @@ impl<F: Field> SubCircuit<F> for SigCircuit<F> {
...
     // so the returned pair is consisted of same values
     fn min_num_rows_block(block: &crate::witness::Block<F>) -> (usize, usize) {
-        let row_num = Self::min_num_rows(block.circuits_params.max_txs);
-        (row_num, row_num)
+        let row_num = Self::min_num_rows();
+        
+        let tx_count = block.txs.len();
+        
+        // TODO: Take the tx limit by block? 
+        let max_tx_count = block.circuits_params.max_txs;
```
Comment (lispc) [inline zkevm-circuits/src/sig_circuit.rs:258]: this is correct.  Sig circuit allocates `calc_required_advices(MAX_NUM_SIG)` advice columns for halo2-libs, which may not be COLUMN_NUM_LIMIT, usually less. `COLUMN_NUM_LIMIT` is a loose limit, usually irrelevant for runtime circuit.   yes the codes here is a bit confusing eg we implicitly require block.circuits_params.max_txs <= MAX_NUM_SIG but that is not documented...

**Comment by lispc** [comment]
the standalone MAX_NUM_SIG constant is tech debt. You can find SuperCircuit is `SuperCircuit::<Fr, MAX_TXS, MAX_CALLDATA,MAX_INNER_BLOCKS,  MOCK_RANDOMNESS>`.   Likewise, SigCircuit should be SigCircuit::<MAX_NUM_SIG>.   Currently we can use MAX_TXS as MAX_NUM_SIG.

**Comment by lispc** [comment]
what is left in this PR before being converted from "DRAFT" to "ready for review"? Do you plan to impl something like `calc_required_advices` for EccCircuit too?   I think it is optional.   Adding some sanity check like `1<<LOG_DEGREE * num_advice > MAX_NUM_SIG * CELLS_PER_SIG` or `1<<LOG_DEGREE * 33 > 50 * 6_900 + 50 * 405_500 + 2 * 6_627_500` somewhere is enough?

**Review by lispc — APPROVED** [review]

**Pull request merged**